lms/services/assignment.py

```diff
@@ -10,55 +13,149 @@ def __init__(self, db):
         self._db = db
 
     @lru_cache(maxsize=128)
-    def get(self, tool_consumer_instance_guid, resource_link_id):
-        return (
-            self._db.query(Assignment)
-            .filter_by(
-                tool_consumer_instance_guid=tool_consumer_instance_guid,
-                resource_link_id=resource_link_id,
-            )
-            .one_or_none()
+    def get(
+        self,
+        tool_consumer_instance_guid,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
+    ):
+        """Get an assignment using using resource_link_id, ext_lti_assignment_id or both."""
+
+        query = self._db.query(Assignment).filter(
+            Assignment.tool_consumer_instance_guid == tool_consumer_instance_guid
         )
+        if resource_link_id and not ext_lti_assignment_id:
+            # Non canvas assignments
+            query = query.filter(Assignment.resource_link_id == resource_link_id)
+        elif not resource_link_id and ext_lti_assignment_id:
+            # creating/editing a canvas assignment
+            query = query.filter(
+                Assignment.ext_lti_assignment_id == ext_lti_assignment_id
+            )
+        elif resource_link_id and ext_lti_assignment_id:
+            query = query.filter(
+                (
+                    # Regular canvas launch
+                    (
+                        (Assignment.resource_link_id == resource_link_id)
+                        & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                    )
+                    # Configuring a file assignment that was stored in the DB
+                    # before all canvas assignments were stored in the DB
+                    | (
+                        (Assignment.resource_link_id == resource_link_id)
+                        & (Assignment.ext_lti_assignment_id.is_(None))
+                    )
+                    #  First launch of a newly configured assignment
+                    | (
+                        (Assignment.resource_link_id.is_(None))
+                        & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                    )
+                )
+            )
+        else:
+            log.exception(
```

Comment:
bddtests expect this behavior, ie not raising here.

Reply:
Hmm. I think we may have conflicting uses of `AssignmentService.get()`. In the LTI launch predicates the code seems to deliberately call `AssignmentService.get()` with just a `resource_link_id` that might be `None` and handles the fact that `get()` might return `None` by calling `bool` on it. The code does that here:

https://github.com/hypothesis/lms/blob/c86dfaecd820fa44e78c25242dfeed78a0a0a75c/lms/views/predicates/_lti_launch.py#L34-L42

and again here:

https://github.com/hypothesis/lms/blob/c86dfaecd820fa44e78c25242dfeed78a0a0a75c/lms/views/predicates/_lti_launch.py#L114-L116

In these cases I don't think you want to log an exception since this appears to be normal, expected behaviour. So an exception in the logs would just be confusing. And logging an exception will report the exception to Sentry as well.

In other places where `get()` is called the code seems to assume that it will *not* return `None`. This happens in the Canvas Files API views:

https://github.com/hypothesis/lms/blob/c86dfaecd820fa44e78c25242dfeed78a0a0a75c/lms/views/api/canvas/files.py#L37-L47

And it also happens in two places in the basic LTI launch views:

https://github.com/hypothesis/lms/blob/c86dfaecd820fa44e78c25242dfeed78a0a0a75c/lms/views/basic_lti_launch.py#L165-L168

https://github.com/hypothesis/lms/blob/c86dfaecd820fa44e78c25242dfeed78a0a0a75c/lms/views/basic_lti_launch.py#L209-L217

I *think* these three are assuming that the return value is never `None`?

So there are two conflicting use cases. So I think you might want to split this into two different methods:

```python
class AssignmentService:
    def get(self, ...):
        # This will raise an exception if both resource_link_id and ext_lti_assignment_id are None.
        # This will also raise an exception if there's no matching assignment.
        # I don't *think* we expect these exceptions to ever actually get raised so we can just
        # allow the app to crash with a 500.

    def exists(self, ...):
        # This one always returns either True or False, never raises.
        # Alternatively this could return an assignment or None. But I think True or False might suffice.
        try:
            self.get(*args, **kwargs)
        except ValueError:
            return False
        else:
            return True
```

Reply:
There are two issues here:

- Not passing any of the ids to `.get()`. This is not expected behavior I think, we should detect that early and issue a 400 or similar. The bddtest fail if I raise here but I think only because they are trying weird cases around the lti spec.

    Based on this I think we don't 4xx on those (missing resource_link_id) because we do some sort of lti redirect. https://github.com/hypothesis/lms/blob/onedrive-auth-page/lms/validation/_lti_launch_params.py#L52 

- Not checking for none on get. The views on basic_lti_launch assume there's a document and don't check for None because they come from the `db_configured` predicate that checks exactly that (https://github.com/hypothesis/lms/blob/c86dfaecd820fa44e78c25242dfeed78a0a0a75c/lms/views/basic_lti_launch.py#L160)

     Then use on the predicate itself, here `get_document_url` is checking two things, that there's a row in the db and that it has a document. Refactoring it with `exists()` will mean doing a query to check if there's an assignment and another to get it and check the document. We could return the assignment on `exists` then but then we really wound't have an use for the `get`, will be just renaming it and between `get -> Assignment` and `exist -> Assignment` I prefer the first one.

Reply:
Yeah it's doing a redirect instead of 404ing in order to be compliant with the LTI spec. It was necessary in order to get LTI certification.

I think the code that's calling `get(resource_link_id=None)` is doing so deliberately. It looks to me like it's doing `resource_link_id = request.params.get("resource_link_id")` instead of just `request.params["resource_link_id"]` deliberately so that `resource_link_id` will be `None` if it's not in the `request.params`. 

Hmm. Ok, so I think it's actually `get_document_url()` rather than `get()` that has two different use-cases?

1. The predicates are calling `get_document_url()`, possibly passing it `resource_link_id=None`, or just a `resource_link_id` that doesn't exist in the DB, and they expect it to return a document URL or `None`. In this case you don't want to raise if `resource_link_id` is `None` or if there's no matching assignment, you just want to return `None`.

2. The views are calling `get_document_url()` and _assuming_ (because the predicates must have already run) that it's going to return a document URL and not `None`. In this case you could raise if `resource_link_id` is `None` or if there's no matching assignment, because this should never happen.

So you might want separate `get_document_url()` and `has_document_url()` methods?

```python
class AssignmentService:
    @lru_cache
    def get(self, ...):
        # Raise if both resource_link_id and ext_lti_assignment_id are None.
        # Raise if there's no matching assignment.

    def get_document_url(self, ...):
        # Calls get() so it will raise if both resource_link_id and ext_lti_assignment_id are None
        # or if there's no matching assignment.
        # Also raise if there is a matching assignment but it has no document_url.

    def has_document_url(self, *args, **kwargs):
        try:
            self.get_document_url(*args, **kwargs)
        except WhateverError:
            return False
        else:
            return True
```

Note that these won't result in multiple DB queries because both `get_document_url()` and `has_document_url()` work by calling `get()` which has `@lru_cache` on it.

The purpose of all this is to allow `get()` and `get_document_url()` to raise if both `resource_link_id` and `ext_lti_assignment_id` are `None` or if there's no matching assignment, because we never expect that to happen. But in the predicates, where we need it to *not* raise, those can call `has_document_url()` instead.

I don't think `get()` can do `log.exception()` as you have it currently because I think that will report exceptions to Sentry during expected behaviour, the edge cases that those BDD tests are covering.

Reply:
The point of all this (in case it's not clear) is to allow the `get()` method to look something like this:

```python
    @lru_cache
    def get(self, tool_consumer_instance_guid, resource_link_id=None, ext_lti_assignment_id=None):
        assert resource_link_id or ext_lti_assignment_id, "Can't get an assignment without a resource_link_id or ext_lti_assignment_id"

        if resource_link_id and ext_lti_assignment_id:
            return self._get_for_canvas_launch(resource_link_id, ext_lti_assignment_id, ...)

        if resource_link_id:
            return self._get_by_resource_link_id(resource_link_id, ...)

        return self._get_by_ext_lti_assignment(ext_lti_assignment_id, ...)
```

You might raise something else rather than an `assert`. But we can make it crash so that we'll get notified if anything is calling `get()` without a `resource_link_id` or an `ext_lti_assignment_id`. At the same time the _predicates_ can use the `has_document_url()` method to avoid any crash